### ISSUE TITLE
Enforce file quota

### ISSUE THREAD

**Issue by devincowan** [hydroshare/hydroshare#5228]
**Describe the feature you'd like and what it will do**
File quota system was implemented circa https://github.com/hydroshare/hydroshare/issues/1566.
However, the quota has not been enforced -- warning notifications are provided but to date users have been permitted to exceed their quota allotments.

Doc used to date: https://docs.google.com/document/d/1OxrX4qPkeYsU0WPu7BMODakCcvSumidallXZjFpnkgo/edit?usp=sharing

We need to turn quota enforcement on.
This will include:
---

### **Policy Decisions**
- [x] Decide policy on incentives and thresholds
- [ ] ~Different quota thresholds for .edu and .gov emails?~
- [x] Published resources will not count toward quota
- [x] How do we handle existing users who are over quota? Example: user currently has 500GB of storage. When we flip the switch and turn on enforcement, should this user be allowed to retain all of these files? We decided to notify current users who are over quota
- [x] Bag “shadow” – if someone has a large resource that they generate a bag for, we don’t count toward quota, but once in the cloud we cuahsi would pay for double the storage. If no downloads within a year, should we "offboard" the resource (remove the bag).

### **HS Development Tasks**
- [x] Form within HS UI Profile to request more quota (including how much, institution, justification).
- [x] Simplified admin approval process in Mezanine.
- [x] Add admin approval process via email
- [x] Add ability to not count published resources against quota
- [x] Add progress bar in the profile page would be very useful to represent this information. It can indicate the current amount of quota used as a percentage of total quota allocated. Extra credit if it is color coded.
- [x] Break down quota by user/data zone
- [x] Users are required to ask for more quota when a new file is uploaded and the quota has been exceeded by the previous upload -- If the upload was rejected then it would be direct for a user to ask for more
- [x] What happens if someone goes over quota (starts grace period) but then reduces their storage? Does the grace period reset? (Looks like the answer is [YES](https://github.com/hydroshare/hydroshare/blob/2.8.1/hs_core/tasks.py#L370))
- [x] Modify the "daily" script. Emails to admin/help should be monthly. User gets notification when login (every time). Admins we will keep with a once-per-month notification. User also gets notified every time that they go over the "soft" limit
- [x] Ensure that we are [enforcing quota at the proper times](https://github.com/hydroshare/hydroshare/issues/5228#issuecomment-1745148581)
- [x] Published resource % should be configurable in the admin dashboard for QuotaMessages
- [x] "Revoke" quota request -> "cancel"
- [x] "File upload disabled due to quota enforcement" -> "Due to your storage being over Quota"
- [x] Check that remaining grace period is reset when UQ below soft limit
- [x] "Current users over quota are contacted and quota is adjusted" -- check email process / management script
- [x] ~#5310~ created a separate issue for this
- [x] Enforcement for uploads conducted via API and HS UI (ideally before the upload starts to prevent user waiting only to find that they have exceeded quota after the upload completes)
- [x] ~#5318~ created a separate issue for this
- [x] Just have default messaging for users that don't have an iRods account
- [x] Uploads in process are completed even if the results in going over the quota limit (unless we know the upload size before upload starts)
- [x] FB: more obvious red notification above the FB
- [x] make JS notifications persistent so that it doesn't dismiss before there's a chance to read
- [x] Check enforcement when copying files from UZ to DZ: [yes, enforced in file_add_pre_process](https://github.com/hydroshare/hydroshare/blob/2.12/irods_browser_app/views.py#L146)
- [x] Write additional tests especially for handling of QuotaException
- [x] Currently, we rely on a celery task to decrement the [remaining_grace_period](https://github.com/hydroshare/hydroshare/blob/2.8.1/hs_core/tasks.py#L331). This is a bad idea. What if celery task stalls? What if the task gets run manually instead of triggered periodically? Instead, we could store the date/time that the grace period will end

### **Devops Tasks**
- [x] Ensure protections exist to prevent issues (undesired notifications etc) on non-prod deploys
- [ ] ~Add the rules that exist on the beta dataZone to the beta userZone~
- [ ] ~After additional testing, create a dot release checklist for deployment of the new engine, msi, and rule on production~

### **Irods Development Tasks**
- [ ] ~Ensure microservice "behaves" (ideally this would include [performance enhancement for multiple file deletion](https://github.com/hydroshare/hydroshare/issues/5235))~ we decided to defer the iRods upgrades until after this issues is released
- [ ] ~#5329 (this could be done with rules or with the internal iRods quota system. However, at this time we have decided to instead just notify support if a user that has already exceeded quota continues to upload to the userZone)~ we have also decided to defer this work. 

### **Messaging and Documentation**
- [ ] User messaging, update help.hydroshare.org
- [ ] Quota messages are stored in the db -- they must be updated
- [ ] Notify current users who are over quota

**Comment by devincowan** [comment]
related to https://github.com/hydroshare/hydroshare/issues/4447

**Comment by devincowan** [comment]
Some places we need to check enforcement:

### file add
https://github.com/hydroshare/hydroshare/blob/2.8.1/hs_core/hydroshare/utils.py#L987
![image](https://github.com/hydroshare/hydroshare/assets/17934193/d8c99263-21ca-4a91-9bb2-87bb9d3dece9)
The formatting is pretty ugly–we should fix it


### Resource create
https://github.com/hydroshare/hydroshare/blob/master/hs_core/hydroshare/utils.py#L832
Currently, if I am over quota and enforce=True, a generic error is shown when I try to create a new resource–even though the create-resource request actually returns the JSON containing quota message:
![image](https://github.com/hydroshare/hydroshare/assets/17934193/8ca2ad0a-9998-4407-b832-1c46117cbf89)


### Resource copy
I don’t think that we validate before starting the copy, we just probably throw an exception during the copying process
![image](https://github.com/hydroshare/hydroshare/assets/17934193/ea4965c0-5e86-4211-bf54-97929b1a91fc)
https://github.com/hydroshare/hydroshare/blob/2.8.1/hs_core/tasks.py#L568

### Resource version
Doesn’t look like we validate before copying
https://github.com/hydroshare/hydroshare/blob/2.8.1/hs_core/tasks.py#L604
This is a problem because [we lock the resource during copy](https://github.com/hydroshare/hydroshare/blob/2.8.1/hs_core/views/__init__.py#L1136), so the resource would remain locked after the QuotaException
We [attempt to unlock it](https://github.com/hydroshare/hydroshare/blob/2.8.1/hs_core/tasks.py#L661) but in this case the `ori_res` will be unbound:
Traceback (most recent call last):
  File "/hydroshare/hs_core/tasks.py", line 661, in create_new_version_resource_task
    ori_res.locked_time = None
UnboundLocalError: local variable 'ori_res' referenced before assignment
Owner transfer
https://github.com/hydroshare/hydroshare/blob/2.8.1/hs_core/models.py#L2340
This is well handled! Other cases should be modeled after this.
Replicate bag or ingest from userzone
https://github.com/hydroshare/hydroshare/blob/2.8.1/hs_core/tasks.py#L695
Add url file to res
https://github.com/hydroshare/hydroshare/blob/master/hs_core/views/resource_folder_hierarchy.py#L553
Looks like this is still possible even after quota exceeded
Unzip (via UI and API)
Tested, looks like this is currently permitted even for user that is over quota.
API file add
Endpoints look like they use the[ resource_file_add_pre_process](https://github.com/hydroshare/hydroshare/blob/2.8.1/hs_core/hydroshare/utils.py#L976) which calls validate_user_quota

Write additional tests especially for handling of QuotaExcepti

**Comment by devincowan** [comment]
@Maurier you had suggested a "progress bar" on the profile page for Quota. Here is what I've implemented. I don't love it so if you have suggestions I'm all ears...thanks!
<img width="582" alt="image" src="https://github.com/hydroshare/hydroshare/assets/17934193/cde7ba92-19df-4968-82e4-64b82655d61f">

**Comment by devincowan** [comment]
updated progress bars with tooltips thanks for the suggestion @Maurier 
<img width="593" alt="image" src="https://github.com/hydroshare/hydroshare/assets/17934193/18062e4c-fda2-4f97-8ccb-2c286b82c8c1">
